exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/BloomFilterCreator.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.work.filter;
+
+import io.netty.buffer.DrillBuf;
+import org.apache.drill.exec.memory.BufferAllocator;
+
+public class BloomFilterCreator {
```

Comment:
Could you pls add a description either here or another appropriate place about how the BF creation works for a generalized multi-column join.. e.g  t1.a1 = t2.a2 AND t1.b1 = t2.b2.   In this case, assume t2 is the build side.  My understanding is as follows: 
  - create a compound hash value using both build side columns: hash(a2, hash(b2))   (this may not be the exact way it is being created but same general idea).
  - add this to the BF, setting appropriate bits to 1
  - on probe side, do the same for a1, b1 and check BF for either false/possibly true

Let me know if this is correct.   This also implies that the size of BF is independent of the number of columns being joined.

Reply:
Current implementation is one bloom filter one join column.  To your example, multi-column join , will generate two bloom filters.  The reason to this implementation is to achieve one vector column memory access by one  hash computation. But the Murmur hash 's complex computation ate up the pipeline performance , the result performance does not so good. So I will change it to your assumes implementation.

Reply:
Hmm...if the 2 columns create separate bloom filters, where does the ANDing (intersection) of the two BFs happen ?  In the current implementation there's an ORing of the BFs but I did not see ANDing (maybe I missed, so let me know).  As far as the memory usage is concerned, we can continue down this path and evaluate both CPU cost and memory usage for multi-column joins where the cardinality of the columns ranges from low to high.

Reply:
There's no intersection of the two separate BFs. The intersection meaning is implemented indirectly at the probe side `ScanBatch`. The separated BFs filter the probe side columns independently and set a BitSet (which indicates the possible rows matched the join condition ) together. This part of codes is at the `ScanBatch`'s `applyRuntimeFilter` and `computeBitSet` methods.

Reply:
Btw, if there's too much multi-column joins , we can add an option to limit the generated  BloomFilter numbers. That's to say there's no need to generate one BF one joint column. We can choose to abandon generating some BFs to save the CPU cost and memory usage.

Reply:
It would be good to add javadoc with this description of multi-column join and how the implicit intersection is done later during BF application in the Scan rather than at the time of BF creation.   This is different from normal hash join where a single hash value is created for multi-column during creation and probing.

Reply:
This description will be add at the `ScanBatch`'s `applyRuntimeFilter `.